ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixParticipant.java

```diff
@@ -221,13 +222,150 @@ public ReplicaSyncUpManager getReplicaSyncUpManager() {
+    }
+    if (newReplicaInfoAdded) {
+      // we update InstanceConfig only when new replica info is added (skip updating if replica is already present)
+      instanceConfig.getRecord().setMapFields(mountPathToDiskInfos);
+      logger.info("Updating config: {} in Helix by adding partition {}", instanceConfig, partitionName);
+      additionResult = helixAdmin.setInstanceConfig(clusterName, instanceName, instanceConfig);
+    }
+    return additionResult;
+  }
+
+  /**
+   * Remove old/existing replica info from {@link InstanceConfig} that associates with current data node.
+   * @param replicaId the {@link ReplicaId} whose info should be removed.
+   * @param instanceConfig {@link InstanceConfig} to update.
+   * @return {@code true} replica info is successfully removed. {@code false} otherwise.
+   */
+  private boolean removeOldReplicaInfo(ReplicaId replicaId, InstanceConfig instanceConfig) {
+    boolean removalResult = true;
+    boolean replicaFound = false;
+    String partitionName = replicaId.getPartitionId().toPathString();
+    Map<String, Map<String, String>> mountPathToDiskInfos = instanceConfig.getRecord().getMapFields();
+    Map<String, String> diskInfo = mountPathToDiskInfos.get(replicaId.getMountPath());
+    if (diskInfo != null) {
+      String replicasStr = diskInfo.get(ClusterMapUtils.REPLICAS_STR);
+      if (!replicasStr.isEmpty()) {
+        List<String> replicaInfos =
+            new ArrayList<>(Arrays.asList(replicasStr.split(ClusterMapUtils.REPLICAS_DELIM_STR)));
+        // if any element is removed, that means old replica is found in replicasStr.
+        replicaFound = replicaInfos.removeIf(
+            info -> (info.split(ClusterMapUtils.REPLICAS_STR_SEPARATOR)[0]).equals(partitionName));
+
+        // We update InstanceConfig only when replica is found in current instanceConfig. (This is to avoid unnecessary
+        // notification traffic due to InstanceConfig change)
+        if (replicaFound) {
+          StringBuilder newReplicasStrBuilder = new StringBuilder();
+          // note that old replica info has been removed from "replicaInfos"
+          for (String replicaInfo : replicaInfos) {
+            newReplicasStrBuilder.append(replicaInfo).append(ClusterMapUtils.REPLICAS_DELIM_STR);
+          }
+          // update diskInfo and MountPathToDisk map
+          diskInfo.put(ClusterMapUtils.REPLICAS_STR, newReplicasStrBuilder.toString());
+          mountPathToDiskInfos.put(replicaId.getMountPath(), diskInfo);
+          // update InstanceConfig
+          instanceConfig.getRecord().setMapFields(mountPathToDiskInfos);
+          logger.info("Updating config: {} in Helix by removing partition {}", instanceConfig, partitionName);
+          removalResult = helixAdmin.setInstanceConfig(clusterName, instanceName, instanceConfig);
+        }
+      }
+    }
+    if (!replicaFound) {
+      logger.warn("Partition {} is not found on instance {}, skipping removing it from InstanceConfig in Helix.",
```

Comment:
here too, should we return false, as we actually didnt find anything to remove?

Reply:
refer to previous explanation. Thanks for pointing it out. It's actually intended behavior in that case.